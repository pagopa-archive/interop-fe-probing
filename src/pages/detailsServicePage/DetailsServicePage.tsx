import { Grid, Typography, Button } from '@mui/material'
import { BarChart } from '../../components/charts/barChart/BarChart'
import { LineChart } from '../../components/charts/lineChart/LineChart'
import ArrowBackIcon from '@mui/icons-material/ArrowBack'
import { InformationBlock } from '../../components/informationBlock/InformationBlock'
import { ChartsLegend } from '../../components/charts/chartsLegend/ChartsLegend'
import { useQuery } from '@tanstack/react-query'
import apiRequests from '../../api/apiRequests'
import { useTranslation } from 'react-i18next'
import stores from '../../store/Store'
import { useNavigate, useParams } from 'react-router-dom'

// elements for the legend component
const legendElements = [
  { label: 'E-service online', color: '#17324D' },
  { label: 'Monitoraggio sospeso', color: '#A2ADB8' },
  { label: 'E-service offline', color: '#FE6666' },
]

const viewInCatalogue = (): void => {
  console.log('view in catalogue')
}

export const DetailsServicePage: React.FC = () => {
  const { t } = useTranslation(['general', 'detailsPage'])

  const navigate = useNavigate()

  const params = useParams()

  const [updateSnackbar] = stores.useSnackbarStore((state) => [state.updateSnackbar])

  const eserviceRecordId = params.id ?? ''

  async function fetchServiceMainData(value: string): Promise<any> {
    return apiRequests.getServiceMainData(value)
  }

  async function fetchServiceProbingData(value: string): Promise<any> {
    return apiRequests.getServiceProbingData(value)
  }

  async function fetchServiceStatisticsData(value: string): Promise<any> {
    return apiRequests.getServiceStatisticsData(value)
  }

  const { data: mainData, isSuccess: mainDataReady } = useQuery({
    queryKey: ['serviceMainData', eserviceRecordId],
    queryFn: () => fetchServiceMainData(eserviceRecordId),
    onError: (error) => updateSnackbar(true, t('errorRequest', { ns: 'general' }), 'error'),
  })

  const {
    data: probingData,
    isSuccess: probingDataReady,
    refetch,
  } = useQuery({
    queryKey: ['serviceProbingData', eserviceRecordId],
    queryFn: () => fetchServiceProbingData(eserviceRecordId),
    onError: (error) => updateSnackbar(true, t('errorRequest', { ns: 'general' }), 'error'),
    keepPreviousData: false,
  })

  const { data: statisticsData, isSuccess: statisticsDataReady } = useQuery({
    queryKey: ['serviceStatisticsData', eserviceRecordId],
    queryFn: () => fetchServiceStatisticsData(eserviceRecordId),
    onError: (error) => updateSnackbar(true, t('errorRequest', { ns: 'general' }), 'error'),
  })

  return (
    <>
      {mainDataReady && probingDataReady && statisticsDataReady ? (
        <Grid container direction="column" spacing={1} sx={{ height: '100%' }}>
          <Grid item sx={{ textAlign: 'center' }} my={5}>
            <Typography variant="h4" component="h1">
              Probing test 3
            </Typography>
            <Typography variant="body1">{t('subtitle', { ns: 'detailsPage' })}</Typography>
          </Grid>
          <Grid item alignSelf={'center'} width={'40%'}>
            <InformationBlock
              mainData={mainData}
              probingData={probingData}
              reloadProbingDetails={refetch}
              viewInCatalogue={viewInCatalogue}
            />
          </Grid>
          <Grid item sx={{ mt: 2 }} flexGrow={1}>
            <Typography
              sx={{
                fontSize: '1.4em',
                fontWeight: 'bold',
                color: '#17324D',
                textAlign: 'center',
              }}
            >
              {t('chartsTitle', { ns: 'detailsPage' })}
            </Typography>
            <Grid item container justifyContent="center" gap={10}>
              <Grid item>
                <LineChart data={statisticsData.values} />
              </Grid>
              <Grid item>
                <Grid container direction="column" rowSpacing={2}>
                  <Grid item>
                    <BarChart data={statisticsData.percentages} />
                  </Grid>
                  <Grid item>
                    <ChartsLegend legendElements={legendElements} />
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
            <Grid
              item
              container
              my={10}
              direction={'row'}
              alignItems={'center'}
              justifyItems={'center'}
              justifyContent={'center'}
            >
              <Button
                onClick={() => navigate(-1)}
                color={'primary'}
                disableRipple
                sx={{
                  fontWeight: 'bold',
                  border: 'none!important',
                  textTransform: 'none',
                  '&:hover': {
                    backgroundColor: 'transparent',
                  },
                }}
                variant="outlined"
                startIcon={<ArrowBackIcon />}
              >
                {t('goBack', { ns: 'detailsPage' })}
              </Button>
            </Grid>
          </Grid>
        </Grid>
      ) : (
        <Typography variant="h4" component="h1">
          Loading
        </Typography>
      )}
    </>
  )
}
