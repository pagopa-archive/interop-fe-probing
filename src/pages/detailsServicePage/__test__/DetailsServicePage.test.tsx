import { describe, expect, test, vi } from "vitest";
import { screen, render, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DetailsServicePage } from "../DetailsServicePage";

const logSpy = vi.spyOn(console, "log");

describe("DettaglioServicePage", () => {
  let container: HTMLElement;

  beforeEach(() => {
    container = render(<DetailsServicePage />).container;
  });

  afterEach(cleanup);

  test("render component", () => {
    expect(container).toMatchSnapshot();
  });

  test("contains titles", async () => {
    const title = screen.getByText(/Probing test 3/i);
    const subtitle = screen.getByText(
      /In questa pagina puoi verificare lo stato di un e-service in tempo reale ed esplorare lo storico delle performance/i
    );
    expect(title).toBeDefined();
    expect(subtitle).toBeDefined();
  });

  test("contains charts", async () => {
    const charts = screen.getAllByRole("img");
    expect(charts).toHaveLength(2);
  });

  test("click Ricarica button", async () => {
    const user = userEvent.setup();
    const button = screen.getByRole("button", { name: "Ricarica" });
    await user.click(button);
    expect(logSpy).toHaveBeenCalledWith("reload info block");
  });

  test("click Visualizza sul catalogo button", async () => {
    const user = userEvent.setup();
    const button = screen.getByRole("button", {
      name: "Visualizza sul catalogo",
    });
    await user.click(button);
    expect(logSpy).toHaveBeenCalledWith("view in catalogue");
  });

  test("click Torna alla lista degli e-service button", async () => {
    const user = userEvent.setup();
    const button = screen.getByRole("button", {
      name: "Torna alla lista degli e-service",
    });
    await user.click(button);
    expect(logSpy).toHaveBeenCalledWith("go back");
  });
});
