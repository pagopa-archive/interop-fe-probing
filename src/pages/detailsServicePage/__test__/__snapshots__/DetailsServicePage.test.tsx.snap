// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DettaglioServicePage > render component 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column css-94xhmw-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column css-13wu608-MuiGrid-root"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-1wm4kle-MuiSkeleton-root"
        style="width: 300px; height: 15px;"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-1wm4kle-MuiSkeleton-root"
        style="width: 600px; height: 15px;"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-g6cc9b-MuiGrid-root"
    >
      <div
        class="MuiStack-root css-19qjufr-MuiStack-root"
      >
        <div
          class="MuiBox-root css-iuqs4u"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-19qjufr-MuiStack-root"
      >
        <div
          class="MuiBox-root css-iuqs4u"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-19qjufr-MuiStack-root"
      >
        <div
          class="MuiBox-root css-iuqs4u"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column css-13wu608-MuiGrid-root"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-1wm4kle-MuiSkeleton-root"
        style="width: 300px; height: 15px;"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-g6cc9b-MuiGrid-root"
    >
      <div
        class="MuiStack-root css-19qjufr-MuiStack-root"
      >
        <div
          class="MuiBox-root css-iuqs4u"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-19qjufr-MuiStack-root"
      >
        <div
          class="MuiBox-root css-iuqs4u"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-19qjufr-MuiStack-root"
      >
        <div
          class="MuiBox-root css-iuqs4u"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column css-1oxcpr4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
        >
          chartsTitle
        </h5>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item css-1quez8l-MuiGrid-root"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-gi27lr-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-naked MuiButton-nakedPrimary MuiButton-sizeSmall MuiButton-nakedSizeSmall MuiButton-root MuiButton-naked MuiButton-nakedPrimary MuiButton-sizeSmall MuiButton-nakedSizeSmall css-14wetfy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        goBack
      </button>
    </div>
  </div>
</div>
`;
